File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if (head is None):
            return False

        slow = head
        fast = head.next

        while(fast is not None):
            slow = slow.next
            fast = fast.next
            if (slow == fast):
                return True
            if (fast == None):
                return False
            fast = fast.next
        return False
        

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head) return nullptr;
        if(!head->next) return head; // could combine with an or?
        ListNode* temp = reverseList(head->next); // traverse first
        head->next->next = head;
        head->next = nullptr; //!! we dont want any forward links
        return temp;
    }
};
